File: analytics/internal/metrics/anomalies.go
package metrics

import (
	"fmt"
	"math"
	"time"

	"analytics/internal/models"
)

// AnomalyConfig contains configuration for anomaly detection.
type AnomalyConfig struct {
	// Spoofing detection
	SpoofingDistanceMultiplier  float64 // How far from mid price (in spread units)
	SpoofingCancelRateThreshold float64 // Cancel rate to trigger detection (0.7 = 70%)
	SpoofingMinOrderSize        float64 // Minimum order size to track

	// Iceberg detection
	IcebergMinFills       int     // Minimum number of fills at same price
	IcebergDepthStability float64 // Max allowed depth change (0.1 = 10%)

	// Flash crash risk detection
	SpreadWideningMultiplier  float64 // Spread increase multiplier to trigger alert
	ThinBookVacuumThreshold   int     // Number of vacuums indicating thin book
	FlowAccelerationThreshold float64 // Net flow acceleration threshold
}

// DefaultAnomalyConfig returns default configuration.
func DefaultAnomalyConfig() *AnomalyConfig {
	return &AnomalyConfig{
		SpoofingDistanceMultiplier:  3.0,
		SpoofingCancelRateThreshold: 0.7,
		SpoofingMinOrderSize:        0,
		IcebergMinFills:             5,
		IcebergDepthStability:       0.1,
		SpreadWideningMultiplier:    2.0,
		ThinBookVacuumThreshold:     3,
		FlowAccelerationThreshold:   -1000.0, // Negative flow acceleration
	}
}

// AnomalyDetector maintains state for anomaly detection.
type AnomalyDetector struct {
	config *AnomalyConfig

	// Spoofing detection state
	trackedOrders       map[string]*trackedOrder
	orderHistory        []orderEvent
	maxOrderHistorySize int

	// Iceberg detection state
	fillHistory        []fillEvent
	maxFillHistorySize int

	// Flash crash detection state
	spreadHistory  []spreadRecord
	flowHistory    []float64
	maxHistorySize int
}

type trackedOrder struct {
	price       float64
	qty         float64
	side        string
	firstSeen   time.Time
	lastSeen    time.Time
	cancelCount int
	updateCount int
}

type orderEvent struct {
	timestamp int64
	price     float64
	qty       float64
	eventType string // "add", "update", "cancel"
	side      string
}

type fillEvent struct {
	timestamp  int64
	price      float64
	volume     float64
	visibleQty float64
}

type spreadRecord struct {
	timestamp int64
	spreadBps float64
}

// NewAnomalyDetector creates a new anomaly detector.
func NewAnomalyDetector(config *AnomalyConfig) *AnomalyDetector {
	if config == nil {
		config = DefaultAnomalyConfig()
	}
	return &AnomalyDetector{
		config:              config,
		trackedOrders:       make(map[string]*trackedOrder),
		orderHistory:        make([]orderEvent, 0, 1000),
		fillHistory:         make([]fillEvent, 0, 1000),
		spreadHistory:       make([]spreadRecord, 0, 1000),
		flowHistory:         make([]float64, 0, 1000),
		maxOrderHistorySize: 1000,
		maxFillHistorySize:  1000,
		maxHistorySize:      1000,
	}
}

// UpdateOrderBook tracks order book changes for spoofing detection.
// T121: Implement large order tracking
func (ad *AnomalyDetector) UpdateOrderBook(timestamp int64, bids, asks []models.PriceQty, midPrice float64) {
	// Track large orders far from mid price
	spreadSize := math.Abs(asks[0].Price - bids[0].Price)
	distanceThreshold := spreadSize * ad.config.SpoofingDistanceMultiplier

	// Track bid orders
	for _, level := range bids {
		distance := midPrice - level.Price
		if distance > distanceThreshold {
			key := makeOrderKey(level.Price, "bid")
			if existing, found := ad.trackedOrders[key]; found {
				existing.lastSeen = time.Now()
				existing.updateCount++
			} else {
				ad.trackedOrders[key] = &trackedOrder{
					price:       level.Price,
					qty:         level.Qty,
					side:        "bid",
					firstSeen:   time.Now(),
					lastSeen:    time.Now(),
					updateCount: 1,
				}
			}
		}
	}

	// Track ask orders
	for _, level := range asks {
		distance := level.Price - midPrice
		if distance > distanceThreshold {
			key := makeOrderKey(level.Price, "ask")
			if existing, found := ad.trackedOrders[key]; found {
				existing.lastSeen = time.Now()
				existing.updateCount++
			} else {
				ad.trackedOrders[key] = &trackedOrder{
					price:       level.Price,
					qty:         level.Qty,
					side:        "ask",
					firstSeen:   time.Now(),
					lastSeen:    time.Now(),
					updateCount: 1,
				}
			}
		}
	}

	// Clean up old tracked orders (older than 30 seconds)
	ad.cleanupTrackedOrders(30 * time.Second)
}

// RecordOrderCancellation tracks cancellations for spoofing detection.
// T122: Implement cancellation rate tracking
func (ad *AnomalyDetector) RecordOrderCancellation(timestamp int64, price float64, side string) {
	key := makeOrderKey(price, side)
	if order, found := ad.trackedOrders[key]; found {
		order.cancelCount++
	}

	ad.orderHistory = append(ad.orderHistory, orderEvent{
		timestamp: timestamp,
		price:     price,
		eventType: "cancel",
		side:      side,
	})

	if len(ad.orderHistory) > ad.maxOrderHistorySize {
		ad.orderHistory = ad.orderHistory[len(ad.orderHistory)-ad.maxOrderHistorySize:]
	}
}

// RecordTradeFill tracks fills for iceberg detection.
// T126: Implement partial fill tracking
func (ad *AnomalyDetector) RecordTradeFill(timestamp int64, price, volume, visibleQty float64) {
	ad.fillHistory = append(ad.fillHistory, fillEvent{
		timestamp:  timestamp,
		price:      price,
		volume:     volume,
		visibleQty: visibleQty,
	})

	if len(ad.fillHistory) > ad.maxFillHistorySize {
		ad.fillHistory = ad.fillHistory[len(ad.fillHistory)-ad.maxFillHistorySize:]
	}
}

// UpdateSpreadHistory tracks spread for flash crash detection.
func (ad *AnomalyDetector) UpdateSpreadHistory(timestamp int64, spreadBps float64) {
	ad.spreadHistory = append(ad.spreadHistory, spreadRecord{
		timestamp: timestamp,
		spreadBps: spreadBps,
	})

	if len(ad.spreadHistory) > ad.maxHistorySize {
		ad.spreadHistory = ad.spreadHistory[len(ad.spreadHistory)-ad.maxHistorySize:]
	}
}

// UpdateFlowHistory tracks net flow for flash crash detection.
func (ad *AnomalyDetector) UpdateFlowHistory(netFlow float64) {
	ad.flowHistory = append(ad.flowHistory, netFlow)

	if len(ad.flowHistory) > ad.maxHistorySize {
		ad.flowHistory = ad.flowHistory[len(ad.flowHistory)-ad.maxHistorySize:]
	}
}

// DetectSpoofing identifies potential spoofing patterns.
// T123: Implement spoofing pattern detection per FR-016
func (ad *AnomalyDetector) DetectSpoofing() []models.Anomaly {
	anomalies := []models.Anomaly{}

	for _, order := range ad.trackedOrders {
		if order.updateCount == 0 {
			continue
		}

		// Calculate cancel rate
		totalEvents := order.updateCount + order.cancelCount
		cancelRate := float64(order.cancelCount) / float64(totalEvents)

		// Check if meets spoofing criteria
		if cancelRate >= ad.config.SpoofingCancelRateThreshold {
			severity := classifySpoofingSeverity(cancelRate, order.cancelCount) // T124
			anomalies = append(anomalies, models.Anomaly{
				Type:     "spoofing",
				Severity: severity,
				Note: fmt.Sprintf("Large %s order at %.2f with %.0f%% cancel rate (%d cancels)",
					order.side, order.price, cancelRate*100, order.cancelCount),
			})
		}
	}

	return anomalies
}

// classifySpoofingSeverity determines spoofing severity.
// T124: Implement severity classification for spoofing
func classifySpoofingSeverity(cancelRate float64, cancelCount int) string {
	if cancelRate >= 0.9 && cancelCount >= 5 {
		return "high"
	} else if cancelRate >= 0.8 && cancelCount >= 3 {
		return "medium"
	}
	return "low"
}

// DetectIcebergOrders identifies potential iceberg order patterns.
// T128: Implement iceberg pattern detection per FR-017
func (ad *AnomalyDetector) DetectIcebergOrders(currentTime int64) []models.Anomaly {
	anomalies := []models.Anomaly{}

	if len(ad.fillHistory) < ad.config.IcebergMinFills {
		return anomalies
	}

	// Group fills by price
	fillsByPrice := make(map[float64][]fillEvent)
	windowStart := currentTime - 300 // 5 minute window

	for _, fill := range ad.fillHistory {
		if fill.timestamp >= windowStart {
			fillsByPrice[fill.price] = append(fillsByPrice[fill.price], fill)
		}
	}

	// Check for iceberg pattern at each price
	for price, fills := range fillsByPrice {
		if len(fills) >= ad.config.IcebergMinFills {
			// T127: Check visible depth stability
			if isDepthStable(fills, ad.config.IcebergDepthStability) {
				anomalies = append(anomalies, models.Anomaly{
					Type:     "iceberg",
					Severity: "medium",
					Note: fmt.Sprintf("Iceberg order detected at %.2f (%d fills with stable visible depth)",
						price, len(fills)),
				})
			}
		}
	}

	return anomalies
}

// isDepthStable checks if visible quantity remains stable across fills.
// T127: Implement visible depth stability monitoring
func isDepthStable(fills []fillEvent, stabilityThreshold float64) bool {
	if len(fills) < 2 {
		return false
	}

	avgQty := 0.0
	for _, fill := range fills {
		avgQty += fill.visibleQty
	}
	avgQty /= float64(len(fills))

	// Check if all fills have similar visible quantity
	for _, fill := range fills {
		deviation := math.Abs(fill.visibleQty-avgQty) / avgQty
		if deviation > stabilityThreshold {
			return false
		}
	}

	return true
}

// DetectFlashCrashRisk identifies conditions indicating flash crash risk.
// T133: Implement flash crash risk pattern detection per FR-018
func (ad *AnomalyDetector) DetectFlashCrashRisk(
	currentSpreadBps float64,
	vacuums []models.LiquidityVacuum,
	currentNetFlow float64,
) []models.Anomaly {
	anomalies := []models.Anomaly{}

	// T130: Check spread widening
	spreadWidening := ad.isSpreadWidening(currentSpreadBps)

	// T131: Check book thinness (number of vacuums)
	thinBook := len(vacuums) >= ad.config.ThinBookVacuumThreshold

	// T132: Check flow acceleration (negative flow getting more negative)
	negativeFlowAccel := ad.isNegativeFlowAccelerating()

	// Combine signals
	signalCount := 0
	if spreadWidening {
		signalCount++
	}
	if thinBook {
		signalCount++
	}
	if negativeFlowAccel {
		signalCount++
	}

	if signalCount >= 2 {
		// At least 2 of 3 signals present
		severity := classifyFlashCrashSeverity(signalCount, vacuums) // T134
		anomalies = append(anomalies, models.Anomaly{
			Type:     "flash_crash_risk",
			Severity: severity,
			Note: fmt.Sprintf("Flash crash risk: widening=%v thin=%v flow=%v",
				spreadWidening, thinBook, negativeFlowAccel),
		})
	}

	return anomalies
}

// isSpreadWidening checks if spread is increasing significantly.
// T130: Implement spread widening detection
func (ad *AnomalyDetector) isSpreadWidening(currentSpreadBps float64) bool {
	if len(ad.spreadHistory) < 10 {
		return false
	}

	// Calculate average spread from recent history
	recentCount := 10
	avgSpread := 0.0
	for i := len(ad.spreadHistory) - recentCount; i < len(ad.spreadHistory); i++ {
		if i >= 0 {
			avgSpread += ad.spreadHistory[i].spreadBps
		}
	}
	avgSpread /= float64(recentCount)

	// Check if current spread is significantly wider
	return currentSpreadBps > avgSpread*ad.config.SpreadWideningMultiplier
}

// isNegativeFlowAccelerating checks if negative flow is accelerating.
// T132: Implement flow velocity tracking
func (ad *AnomalyDetector) isNegativeFlowAccelerating() bool {
	if len(ad.flowHistory) < 5 {
		return false
	}

	// Calculate flow acceleration (change in flow rate)
	recentFlows := ad.flowHistory[len(ad.flowHistory)-5:]

	// Check if flow is consistently negative and getting more negative
	allNegative := true
	acceleration := 0.0

	for i := 1; i < len(recentFlows); i++ {
		if recentFlows[i] >= 0 {
			allNegative = false
		}
		acceleration += recentFlows[i] - recentFlows[i-1]
	}

	return allNegative && acceleration < ad.config.FlowAccelerationThreshold
}

// classifyFlashCrashSeverity determines flash crash risk severity.
// T134: Implement severity classification for flash crash risk
func classifyFlashCrashSeverity(signalCount int, vacuums []models.LiquidityVacuum) string {
	// Count high severity vacuums
	highSeverityVacuums := 0
	for _, v := range vacuums {
		if v.Severity == "high" {
			highSeverityVacuums++
		}
	}

	if signalCount == 3 || highSeverityVacuums >= 3 {
		return "high"
	} else if signalCount == 2 || highSeverityVacuums >= 2 {
		return "medium"
	}
	return "low"
}

// DetectAnomalies performs complete anomaly detection.
// T136: Integrate anomaly detection (spoofing, iceberg, flash crash)
func (ad *AnomalyDetector) DetectAnomalies(
	currentTime int64,
	currentSpreadBps float64,
	vacuums []models.LiquidityVacuum,
	currentNetFlow float64,
) []models.Anomaly {
	anomalies := []models.Anomaly{}

	// Detect spoofing
	spoofingAnomalies := ad.DetectSpoofing()
	anomalies = append(anomalies, spoofingAnomalies...)

	// Detect iceberg orders
	icebergAnomalies := ad.DetectIcebergOrders(currentTime)
	anomalies = append(anomalies, icebergAnomalies...)

	// Detect flash crash risk
	flashCrashAnomalies := ad.DetectFlashCrashRisk(currentSpreadBps, vacuums, currentNetFlow)
	anomalies = append(anomalies, flashCrashAnomalies...)

	return anomalies
}

// cleanupTrackedOrders removes orders that haven't been seen recently.
func (ad *AnomalyDetector) cleanupTrackedOrders(maxAge time.Duration) {
	now := time.Now()
	for key, order := range ad.trackedOrders {
		if now.Sub(order.lastSeen) > maxAge {
			delete(ad.trackedOrders, key)
		}
	}
}

// makeOrderKey creates a unique key for tracking orders.
func makeOrderKey(price float64, side string) string {
	return fmt.Sprintf("%s_%.8f", side, price)
}

File: producer/src/calculators/anomalies.py
"""Anomaly detection for market microstructure analysis.

Detects spoofing, iceberg orders, and flash crash risk signals.
"""
import numpy as np
from typing import Optional
from datetime import datetime, timezone, timedelta
from src.state.symbol_state import TradeTick, OrderBookL2, PriceQty


def detect_spoofing(
    order_book: OrderBookL2,
    mid_price: float,
    cancel_rate_threshold: float = 0.70,
    distance_threshold_bps: int = 50
) -> list[dict]:
    """Detect potential spoofing activity.

    Spoofing is characterized by large far-from-mid orders with high cancel rates.
    Since we don't track individual order lifecycle, we use proxy signals:
    - Large orders far from mid price (>50 bps)
    - Sudden appearance/disappearance of large orders

    Args:
        order_book: Current order book state
        mid_price: Current mid price
        cancel_rate_threshold: Cancel rate threshold (default 70%)
        distance_threshold_bps: Minimum distance from mid in basis points

    Returns:
        List of detected spoofing signals:
        [{
            "type": "spoofing",
            "side": "bid" | "ask",
            "price": 43100.0,
            "quantity": 25.5,
            "distance_bps": 75,
            "severity": "high" | "medium" | "low",
            "note": "Large order far from mid, potential spoofing"
        }]
    """
    anomalies = []

    # T066: Check for large far-from-mid orders on bid side
    for price, qty in order_book.top_bids[:10]:  # Check top 10 levels
        distance_bps = abs((price - mid_price) / mid_price * 10000)

        if distance_bps > distance_threshold_bps:
            # Calculate average quantity for comparison
            avg_qty = np.mean([q for _, q in order_book.top_bids]) if order_book.top_bids else 0

            if qty > avg_qty * 2:  # 2x average is suspicious
                # Severity based on size and distance
                if qty > avg_qty * 5 and distance_bps > 100:
                    severity = "high"
                elif qty > avg_qty * 3:
                    severity = "medium"
                else:
                    severity = "low"

                anomalies.append({
                    "type": "spoofing",
                    "side": "bid",
                    "price": float(price),
                    "quantity": float(qty),
                    "distance_bps": int(distance_bps),
                    "severity": severity,
                    "note": f"Large bid {qty:.2f} at {distance_bps:.0f}bps from mid, potential spoofing"
                })

    # T066: Check for large far-from-mid orders on ask side
    for price, qty in order_book.top_asks[:10]:
        distance_bps = abs((price - mid_price) / mid_price * 10000)

        if distance_bps > distance_threshold_bps:
            avg_qty = np.mean([q for _, q in order_book.top_asks]) if order_book.top_asks else 0

            if qty > avg_qty * 2:
                if qty > avg_qty * 5 and distance_bps > 100:
                    severity = "high"
                elif qty > avg_qty * 3:
                    severity = "medium"
                else:
                    severity = "low"

                anomalies.append({
                    "type": "spoofing",
                    "side": "ask",
                    "price": float(price),
                    "quantity": float(qty),
                    "distance_bps": int(distance_bps),
                    "severity": severity,
                    "note": f"Large ask {qty:.2f} at {distance_bps:.0f}bps from mid, potential spoofing"
                })

    return anomalies


def detect_iceberg(
    trades: list[TradeTick],
    order_book: OrderBookL2,
    price_tolerance_pct: float = 0.10
) -> list[dict]:
    """Detect potential iceberg orders.

    Iceberg orders show: ≥5 fills at same price with stable visible depth (±10%).

    Args:
        trades: Recent trade ticks (recommend 30s window)
        order_book: Current order book state
        price_tolerance_pct: Price tolerance for "same price" (default 0.10%)

    Returns:
        List of detected iceberg signals:
        [{
            "type": "iceberg",
            "side": "bid" | "ask",
            "price": 43250.5,
            "fill_count": 8,
            "total_volume": 45.5,
            "severity": "high" | "medium" | "low",
            "note": "8 fills at same price with stable depth, potential iceberg"
        }]
    """
    anomalies = []

    if len(trades) < 5:
        return anomalies

    # T067: Group trades by price (within tolerance)
    price_groups = {}
    for trade in trades:
        # Find price bucket (rounded to tolerance)
        price_key = round(trade.price / (trade.price * price_tolerance_pct / 100)) * (trade.price * price_tolerance_pct / 100)

        if price_key not in price_groups:
            price_groups[price_key] = {
                "trades": [],
                "total_volume": 0,
                "buy_count": 0,
                "sell_count": 0
            }

        price_groups[price_key]["trades"].append(trade)
        price_groups[price_key]["total_volume"] += trade.volume

        if trade.aggressor_side == "BUY":
            price_groups[price_key]["buy_count"] += 1
        else:
            price_groups[price_key]["sell_count"] += 1

    # T067: Check for iceberg pattern (≥5 fills at same price)
    for price_key, group in price_groups.items():
        fill_count = len(group["trades"])

        if fill_count >= 5:
            # Determine dominant side
            if group["buy_count"] > group["sell_count"]:
                side = "ask"  # Buyers hitting asks = iceberg on ask side
            else:
                side = "bid"  # Sellers hitting bids = iceberg on bid side

            # T069: Severity classification
            if fill_count >= 20:
                severity = "high"
            elif fill_count >= 10:
                severity = "medium"
            else:
                severity = "low"

            anomalies.append({
                "type": "iceberg",
                "side": side,
                "price": float(price_key),
                "fill_count": fill_count,
                "total_volume": float(group["total_volume"]),
                "severity": severity,
                "note": f"{fill_count} fills at ~{price_key:.2f} with stable depth, potential iceberg"
            })

    return anomalies


def detect_flash_crash_risk(
    spread_bps: float,
    depth_imbalance: float,
    flow_acceleration: float,
    spread_threshold_bps: float = 20.0,
    imbalance_threshold: float = 0.3,
    flow_threshold: float = -100.0
) -> Optional[dict]:
    """Detect flash crash risk conditions.

    Flash crash risk present when ≥2 of 3 signals trigger:
    1. Spread widening (>20 bps)
    2. Thin book (imbalance >30%)
    3. Negative flow acceleration (<-100 orders/sec²)

    Args:
        spread_bps: Current spread in basis points
        depth_imbalance: Order book imbalance (-1 to 1)
        flow_acceleration: Rate of change of order flow (orders/sec²)
        spread_threshold_bps: Spread widening threshold
        imbalance_threshold: Imbalance threshold (abs value)
        flow_threshold: Flow acceleration threshold

    Returns:
        Flash crash risk signal or None:
        {
            "type": "flash_crash_risk",
            "triggered_signals": ["spread_widening", "thin_book", "negative_flow"],
            "severity": "high" | "medium" | "low",
            "note": "3 of 3 flash crash signals active"
        }
    """
    # T068: Check each signal
    signals_triggered = []

    # Signal 1: Spread widening
    if spread_bps > spread_threshold_bps:
        signals_triggered.append("spread_widening")

    # Signal 2: Thin book (high imbalance)
    if abs(depth_imbalance) > imbalance_threshold:
        signals_triggered.append("thin_book")

    # Signal 3: Negative flow acceleration
    if flow_acceleration < flow_threshold:
        signals_triggered.append("negative_flow")

    # T068: Require ≥2 of 3 signals
    if len(signals_triggered) < 2:
        return None

    # T069: Severity based on number of signals
    if len(signals_triggered) == 3:
        severity = "high"
    elif len(signals_triggered) == 2:
        severity = "medium"
    else:
        severity = "low"

    return {
        "type": "flash_crash_risk",
        "triggered_signals": signals_triggered,
        "severity": severity,
        "note": f"{len(signals_triggered)} of 3 flash crash signals active",
        "details": {
            "spread_bps": float(spread_bps),
            "depth_imbalance": float(depth_imbalance),
            "flow_acceleration": float(flow_acceleration)
        }
    }


def calculate_flow_acceleration(
    trades: list[TradeTick],
    window_sec: int = 10
) -> float:
    """Calculate flow acceleration (rate of change of order flow).

    Args:
        trades: Trade history
        window_sec: Time window for calculation

    Returns:
        Flow acceleration in orders/sec² (positive = accelerating, negative = decelerating)
    """
    if len(trades) < 2:
        return 0.0

    # Split window into two halves
    now = datetime.now(timezone.utc)
    half_window = timedelta(seconds=window_sec / 2)

    recent_trades = [t for t in trades if (now - t.timestamp) <= half_window]
    older_trades = [t for t in trades if half_window < (now - t.timestamp) <= timedelta(seconds=window_sec)]

    if not recent_trades or not older_trades:
        return 0.0

    # Calculate orders per second for each half
    recent_rate = len(recent_trades) / (window_sec / 2)
    older_rate = len(older_trades) / (window_sec / 2)

    # Acceleration = change in rate / time
    acceleration = (recent_rate - older_rate) / (window_sec / 2)

    return float(acceleration)
